torch/csrc/jit/passes/autocast.cpp

```diff
@@ -0,0 +1,456 @@
+      case aten::addmm:
+      case aten::addmv:
+      case aten::addr:
+      case aten::matmul:
+      case aten::mm:
+      case aten::mv:
+      case aten::linear:
+      case aten::addbmm:
+      case aten::baddbmm:
+      case aten::bmm:
+      case aten::chain_matmul:
+      case aten::_thnn_fused_lstm_cell:
+      case aten::_thnn_fused_gru_cell:
+      case aten::lstm_cell:
+      case aten::gru_cell:
+      case aten::rnn_tanh_cell:
+      case aten::rnn_relu_cell:
+        if (!node->schema().is_mutable()) {
+          castTensorInputs(
+              node, aten::_autocast_to_reduced_precision, current_state());
+        }
+        break;
+
+      // CastPolicy::fp32 (cast all inputs to float32)
+      case aten::native_layer_norm:
+      case aten::acos:
+      case aten::asin:
+      case aten::cosh:
+      case aten::erfinv:
+      case aten::exp:
+      case aten::expm1:
+      case aten::log:
+      case aten::log10:
+      case aten::log2:
+      case aten::log1p:
+      case aten::reciprocal:
+      case aten::rsqrt:
+      case aten::sinh:
+      case aten::tan:
+      case aten::pow:
+      case aten::softplus:
+      case aten::gelu:
+      case aten::layer_norm:
+      case aten::group_norm:
+      case aten::frobenius_norm:
+      case aten::nuclear_norm:
+      case aten::cosine_similarity:
+      case aten::cosine_embedding_loss:
+      case aten::nll_loss:
+      case aten::nll_loss2d:
+      case aten::hinge_embedding_loss:
+      case aten::kl_div:
+      case aten::l1_loss:
+      case aten::smooth_l1_loss:
+      case aten::mse_loss:
+      case aten::margin_ranking_loss:
+      case aten::multilabel_margin_loss:
+      case aten::soft_margin_loss:
+      case aten::triplet_margin_loss:
+      case aten::multi_margin_loss:
+      case aten::binary_cross_entropy_with_logits:
+      case aten::dist:
+      case aten::pdist:
+      case aten::cdist:
+      case aten::renorm:
+        if (!node->schema().is_mutable()) {
+          castTensorInputs(
+              node, aten::_autocast_to_full_precision, current_state());
+        }
+        break;
+
+      // CastPolicy::fp32_set_opt_dtype
+      case aten::prod:
+      case aten::softmax:
+      case aten::log_softmax:
+      case aten::cumprod:
+      case aten::cumsum:
+      case aten::sum:
+        if (!node->schema().is_mutable() && !hasExplicitDtypeArgument(node)) {
+          castTensorInputs(
+              node, aten::_autocast_to_full_precision, current_state());
+        }
+        break;
+
+      // CastPolicy::promote (promote inputs to the widest type)
+      case aten::addcdiv:
+      case aten::addcmul:
+      case aten::atan2:
+      case aten::bilinear:
+      case aten::cat:
+      case aten::_cat:
+      case aten::cross:
+      case aten::dot:
+      case aten::equal:
+      case aten::index_put:
+      case aten::stack:
+      case aten::tensordot:
+      // add, sub, mul, div were added to autocast jit, because aten implicit
```

Comment:
sorry, mind saying a bit more on this? what does this mean?

Reply:
Sure will do.

This is just what autodiff/autograd does differently on things going backward and we've got into mismatched error without explicitly casting inputs. I'll put more context in comments.